### PR TITLE
Fields deriver for JS types and conversions

### LINKED ISSUE #10553: Auto-generate JS <-> OCaml conversion
Plan:
* Create a new deriver `js_type_layout` which builds up a JSON-like description of _the OCaml type_. For a record
```ocaml
{a: string, b: int list}
```
this could derive a JSON object like
```js
{
  type: "record",
  layout: [
    ["a", {type: "string"}],
    ["b", {type: "array", inner: {type: "number"}}]
  ]
}
```

This type description object can then easily used by generic JS code to
* figure out a suitable JS representation of this type,
* convert back and forth between the JS and OCaml representation of a value of this type
* auto-generate a TypeScript type declaration
* ...and probably more

### PR DESCRIPTION
**Pull request by mitschabaude**
* Closes #10553 
* More info in the snarkyjs companion: https://github.com/o1-labs/snarkyjs/pull/158

### CODE REVIEW

File: scripts/build-snarkyjs-node.sh [hunk 1]
```diff
@@ -2,7 +2,12 @@ SNARKY_JS_PATH=$1
 [ -z "$SNARKY_JS_PATH" ] && SNARKY_JS_PATH=src/lib/snarky_js_bindings/snarkyjs
 
 dune b src/lib/crypto/kimchi_bindings/js/node_js \
-&& dune b src/lib/snarky_js_bindings/snarky_js_node.bc.js || exit 1
+&& dune b src/lib/snarky_js_bindings/snarky_js_node.bc.js src/lib/snarky_js_bindings/snarky_js_types.exe || exit 1
+
+# auto-generate js types from field derivers
+# todo declare these files as a dune target somehow
```
Comment (mrmr1993) [inline scripts/build-snarkyjs-node.sh:8]: See the rule in e.g. `/dune` for `graphql_schema.json` for how this is done

File: src/lib/mina_base/account_id.mli [hunk 2]
```diff
@@ -13,6 +13,8 @@ module Digest : sig
 
   val of_field : Snark_params.Tick.Field.t -> t
 
+  val to_field : t -> Snark_params.Tick.Field.t
```
Comment (mrmr1993) [inline src/lib/mina_base/account_id.mli:16]: `to_field_unsafe`, the idea is that this gives type safety for our OCaml code

File: src/lib/fields_derivers_zkapps/fields_derivers_js.ml [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+        ; ("inner", inner)
+        ] ;
+    obj
+
+  let wrapped x obj =
+    obj#js_layout := !(x#js_layout) ;
+    obj
+
+  let with_checked ~name (x : _ Input.t) (obj : _ Input.t) =
+    (* TODO: how to tell the type checker that obj#js_layout is of variant `Assoc? *)
```
Comment (mrmr1993) [inline src/lib/fields_derivers_zkapps/fields_derivers_js.ml:112]: Change the `[>` in the type to `[`

File: scripts/build-snarkyjs-node.sh (hunk 1 above)
Comment (mrmr1993) [reply scripts/build-snarkyjs-node.sh:8]: Like with that, it would be good to use promotion so that we can see the results in-tree

**Review by mrmr1993 — APPROVED** [review]
This is fab

File: scripts/build-snarkyjs-node.sh (hunk 1 above)
Comment (mitschabaude) [reply scripts/build-snarkyjs-node.sh]: thanks - I made this work now!
https://github.com/MinaProtocol/mina/pull/10938/commits/fee40045a44a47788e437c13baee2208f71800fa
https://github.com/o1-labs/snarkyjs/pull/158/commits/1b287823b09d6a49f6ff764932a1d70007787a20

**Review by bkase — APPROVED** [review]
Looks great!

**Pull request merged**